apps/aecore/lib/aecore/account/tx/spend_tx.ex

```diff
@@ -54,88 +57,87 @@ defmodule Aecore.Account.Tx.SpendTx do
   @doc """
   Checks wether the amount that is send is not a negative number
   """
-  @spec is_valid?(SpendTx.t()) :: boolean()
-  def is_valid?(%SpendTx{receiver: receiver, amount: amount}) do
+  @spec is_valid?(SpendTx.t(), DataTx.t()) :: boolean()
+  def is_valid?(%SpendTx{receiver: receiver} = tx, data_tx) do
+    senders = DataTx.senders(data_tx)
+
     cond do
-      amount < 0 ->
+      tx.amount < 0 ->
         Logger.error("The amount cannot be a negative number")
         false
 
+      tx.version != get_tx_version() ->
+        Logger.error("Invalid version")
+        false
+
       !Wallet.key_size_valid?(receiver) ->
         Logger.error("Wrong receiver key size")
         false
 
+      length(senders) != 1 ->
+        Logger.error("Invalid senders number")
+        false
+
       true ->
         true
     end
   end
 
-  @doc """
-  Makes a rewarding SpendTx (coinbase tx) for the miner that mined the next block
-  """
-  @spec reward(SpendTx.t(), Account.t()) :: Account.t()
-  def reward(%SpendTx{} = tx, account_state) do
-    Account.transaction_in(account_state, tx.amount)
-  end
-
   @doc """
   Changes the account state (balance) of the sender and receiver.
   """
   @spec process_chainstate!(
-          SpendTx.t(),
-          Wallet.pubkey(),
-          non_neg_integer(),
-          non_neg_integer(),
+          ChainState.account(),
+          tx_type_state(),
           non_neg_integer(),
-          AccountStateTree.tree(),
-          tx_type_state()
-        ) :: {AccountStateTree.tree(), tx_type_state()} | {:error, String.t()}
-  def process_chainstate!(%SpendTx{} = tx, sender, fee, nonce, block_height, accounts, %{}) do
-    sender_account_state = Account.get_account_state(accounts, sender)
-
-    case preprocess_check!(tx, sender, sender_account_state, fee, block_height, nonce, %{}) do
-      :ok ->
-        new_sender_account_state =
-          sender_account_state
-          |> deduct_fee(fee)
-          |> Account.transaction_out(tx.amount * -1, nonce)
-
-        new_accounts = AccountStateTree.put(accounts, sender, new_sender_account_state)
-        receiver = Account.get_account_state(accounts, tx.receiver)
-        new_receiver_acc_state = Account.transaction_in(receiver, tx.amount)
-
-        {AccountStateTree.put(new_accounts, tx.receiver, new_receiver_acc_state), %{}}
-
-      {:error, _reason} = err ->
-        throw(err)
-    end
+          SpendTx.t(),
+          DataTx.t()
+        ) :: {ChainState.accounts(), tx_type_state()}
+  def process_chainstate!(accounts, %{}, block_height, %SpendTx{} = tx, data_tx) do
+    sender = DataTx.sender(data_tx)
+
+    new_accounts =
+      accounts
+      |> AccountStateTree.update(sender, fn acc ->
+        Account.transaction_in!(acc, block_height, tx.amount * -1)
```

Comment:
I see that this is why `transaction_out` is removed, but we should either keep `transaction_out` and have the `* -1` there, or change the name to something more generic like `apply_transaction`.

Reply:
I will change the name.

Reply:
this solves my https://github.com/aeternity/elixir-node/pull/355#discussion_r182962819 as well